Fix #8019: Adds abbreviated topic name and thumbnail fields in…

kevintab95 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #8019: Adds abbreviated topic name and thumbnail fields in the topic editor

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Commit:
adds abbreviated topic name and topic thumbnail fields

Commit:
add tests

Commit:
fix errors

Commit:
fix lint errors

Commit:
fix html lint issues

Commit:
fix html issue

Commit:
fix html lint issue

kevintab95 commented:
I'm working on test failures. This PR is not ready for review yet.

Commit:
fix failures

kevintab95 commented:
Hi @kevinlee12, can you please take a look at the logs for lint tests? Looks like there's a linting error but the script does not seem to report what the error is or where it might be occurring.

Commit:
fit -> it

kevintab95 commented:
@kevinlee12, it was due to `fit` not being reported by the linting script.

seanlip commented:
@kevintab95 -- one note, you might want to try to get the diff-coverage up to 100%. Backend test coverage should always be at 100% before PRs are merged. (Make sure to test any new behaviour.)

Commit:
add tests for topic_domain andtopic_services changes

Commit:
fix lint issue

kevintab95 commented:
Hi @seanlip, I've added some tests and the diff-coverage is now 100%. PTAL thanks!

seanlip commented on feconf.py:
```diff
@@ -294,6 +294,10 @@
+# Default abbreviated name for a newly-minted topic.
+DEFAULT_ABBREVIATED_TOPIC_NAME = ''
+# Default thumbnail data url for a newly-minted topic.
+DEFAULT_THUMBNAIL_DATA_URL = ''
```
This should be None, and it doesn't need to be defined as a constant here. None means "there's no URL yet"; a string means that it's a valid URL.

kevintab95 replied:
Removed this.

kevintab95 replied:
Done.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -51,6 +51,10 @@ class TopicModel(base_models.VersionedModel):
     # The canonical name of the topic, created by making `name` lowercase.
     canonical_name = ndb.StringProperty(required=True, indexed=True)
+    # The abbreviated name of the topic.
+    abbreviated_name = ndb.StringProperty(indexed=False)
```
indexed=True, required=True

kevintab95 replied:
Done.

kevintab95 replied:
Done.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -51,6 +51,10 @@ class TopicModel(base_models.VersionedModel):
+    # The abbreviated name of the topic.
+    abbreviated_name = ndb.StringProperty(indexed=False)
+    # Topic thumbnail as a dataURI string.
+    thumbnail_data_url = ndb.TextProperty(default=None, indexed=False)
```
Actually, I think it is better to store this in GCS, since then the image can be cached. Otherwise it needs to be reconstituted every time.

@aks681, can you suggest the filepath for where it should go, so that the image is stored separately from any other topic images?

Then we don't need this field, since we can just look up the image based on topic ID. (Sorry for not realizing this earlier.)

aks681 replied:
Maybe, you can use `/topic/thumbnails/<[topicId]>`?
We are already using `/topic/<[topicId]>` to store the RTE images in a topic, so you can use the same functions, just with a different URL.

seanlip replied:
@aks681 Technically this doesn't work so well if the topic ID is "thumbnails". Which generally can't happen but it feels a bit odd to structure it that way.

Are the RTE images for that topic stored in an images subfolder or in the topic root folder? If the former then making something like `/topic/<[topicId]>/thumbnails` should work, and if the latter then we could either (i) change it to the former, or (ii) just store all images in `/topic/<[topicId]>`.

Also, do you think we should be creating a folder and storing just the filename of the thumbnail image in the model? I guess that's probably better than specifying an explicit filepath since the filepath would end up fixing the format of the image (e.g. the thumbnail would always be `/topic/<[topicId]>/thumbnail.png` but we might not want to fix it as png). It would also give us the benefit of being able to keep previous thumbnail images in case we wanted to switch back to them later.

aks681 replied:
The image is stored inside /topic/topicId/assets/image/<[url]>. Yeah, making something like /assets/thumbnails should be fine.
That's what is done now. We just store the filename of the image in the RTE content, and using the context, the full url is generated. Something similar can be done here.

kevintab95 replied:
Okay, I'll make the changes to store thumbnails in `/topic/<topic-id>/assets/thumbnails`. Thanks!

kevintab95 replied:
Done.

seanlip reviewed (changes requested):
Thanks, @kevintab95! I need to defer to @aks681 for a full review on this, but I did take a look at the codeowner files and have a couple of comments.

aks681 commented on core/domain/topic_domain.py:
```diff
@@ -744,20 +756,24 @@ def validate(self):
-    def create_default_topic(cls, topic_id, name):
+    def create_default_topic(
+            cls, topic_id, name,
+            abbreviated_name=feconf.DEFAULT_ABBREVIATED_TOPIC_NAME):
```
Isn't `abbreviated_name` compulsory to create a topic? If so, add the relevant checks in `topic.validate()` verifying that it's non-empty.  Also, check if its length is less than a threshold (@seanlip What should be this threshold?).

Do the other normal checks as well, in `validate`, verifying abbreviated_topic_name and thumbnail_url are strings.
 
Also, I'd recommend not using default arguments here, since all topic objects should have an abbreviated_name and it's specific to that topic, doesn't make sense to have a common one for all created topics. Was there any particular reason why you used it?

seanlip replied:
The threshold should be 12 characters. I agree that there should not be a default name.

kevintab95 replied:
Done.

kevintab95 replied:
Done.

aks681 commented on core/domain/topic_domain.py:
```diff
@@ -430,6 +440,8 @@ def to_dict(self):
         return {
             'id': self.id,
             'name': self.name,
+            'abbreviated_name': self.abbreviated_name,
```
Have you updated the `from_dict` function as well? I can't see that change.

kevintab95 replied:
Added from_dict.

kevintab95 replied:
Added from_dict.

aks681 commented on core/domain/topic_fetchers_test.py:
```diff
@@ -49,7 +49,7 @@ def setUp(self):
         })]
         self.save_new_topic(
-            self.TOPIC_ID, self.user_id, 'Name', 'Description',
+            self.TOPIC_ID, self.user_id, 'Name', '', '', 'Description',
```
Maybe initialize the new fields with some value here? Since, this is a topic specific test.

kevintab95 replied:
Done.

kevintab95 replied:
Done.

aks681 commented on core/templates/dev/head/components/entity-creation-services/topic-creation.service.ts:
```diff
@@ -55,26 +59,28 @@ angular.module('oppia').factory('TopicCreationService', [
-        modalInstance.result.then(function(topicName) {
-          if (topicName === '') {
+        modalInstance.result.then(function(topic) {
+          if (topic.topicName === '') {
```
Isn't `abbreviated_topic_name` also compulsory? If not, maybe we should have a mechanism, where if the topic name  is short, automatically copy that into abbreviated_topic_name.
@seanlip Do you have any thoughts about this?

seanlip replied:
I think it should be compulsory. Adding an abbreviated name is not hard (unlike, say, creating a topic thumbnail image).

kevintab95 replied:
Done.

kevintab95 replied:
Done.

aks681 commented on core/templates/dev/head/domain/topic/TopicObjectFactory.ts:
```diff
@@ -90,6 +109,10 @@ angular.module('oppia').factory('TopicObjectFactory', [
         issues.push('Topic name should not be empty.');
       }
 
+      if (this._thumbnailDataUrl === '') {
```
What about for `_abbreviatedTopicName`?
Also, isn't this just for publishing? Or is this function only checked before publishing?

kevintab95 replied:
Added validaiton.

aks681 commented on core/templates/dev/head/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.ts:
```diff
@@ -58,6 +63,65 @@ angular.module('oppia').directive('topicEditorTab', [
 
+          $scope.getStaticImageUrl = UrlInterpolationService.getStaticImageUrl;
+
+          $scope.showEditThumbnailModal = function() {
```
Disable this when user has no edit access.

kevintab95 replied:
Done.

aks681 reviewed (changes requested):
Did a full pass. Can you also attach screenshots of how the editor looks now?

seanlip commented on core/controllers/topics_and_skills_dashboard.py:
```diff
@@ -132,10 +132,11 @@ class NewTopicHandler(base.BaseHandler):
         name = self.payload.get('name')
-
+        abbreviated_name = self.payload.get('abbreviated_name')
         topic_domain.Topic.require_valid_name(name)
```
Add validation for the abbreviated name as well. (Also add tests to test the validation, missing keys, etc. -- make sure the logic here catches _all_ possible issues with data submitted to the URL, since we should regard it as completely untrusted.)

kevintab95 replied:
Done. Added tests to test the validation in topic_domain_test.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -827,6 +843,24 @@ def update_name(self, new_name):
+        """Updates the thumbnail_data_url of a topic object.
+
+        Args:
+            new_thumbnail_data_url: str. The updated thumbnail_data_url
```
Could also be None right? (If the image gets removed.)

kevintab95 replied:
Done.

kevintab95 replied:
Done.

seanlip reviewed (changes requested):
(Took a fuller pass, and left more comments!)

Commit:
address comments

Commit:
fix tests

Commit:
Merge remote-tracking branch 'upstream/develop' into abbreviated-topic-name-and-thumbnail

Commit:
add eol

Commit:
fix test

Commit:
address review comments

Commit:
add tests for abbrev name validation

kevintab95 commented:
@seanlip and @aks681 I've addressed the review comments. PTAL, thanks!

Here are some screenshots of how the editor looks:
![topicModal](https://user-images.githubusercontent.com/11008603/69748270-fbb0d900-116d-11ea-99fe-0cb4150e9789.png)

![thumbnailEditor](https://user-images.githubusercontent.com/11008603/69748279-ff446000-116d-11ea-938d-89667cafd042.png)

![withanimage](https://user-images.githubusercontent.com/11008603/69748296-023f5080-116e-11ea-8d67-4fd0dba2560e.png)

![cannotpublish](https://user-images.githubusercontent.com/11008603/69748305-07040480-116e-11ea-8bb3-9350ad51215d.png)

Commit:
fix backend and e2e failures

seanlip commented on core/controllers/resources.py:
```diff
@@ -52,7 +52,7 @@ class AssetDevHandler(base.BaseHandler):
     """
 
-    _SUPPORTED_TYPES = ['image', 'audio']
+    _SUPPORTED_TYPES = ['image', 'audio', 'thumbnails']
```
Should it be "thumbnail" to follow the other nouns which are singular?

kevintab95 replied:
Done.

seanlip commented on core/domain/topic_domain.py:
```diff
@@ -827,6 +889,24 @@ def update_name(self, new_name):
+        """
+        self.abbreviated_name = new_abbreviated_name
+
+    def update_thumbnail(self, new_thumbnail):
```
Here and elsewhere I think we should be calling the "thing" thumbnail_filename rather than thumbnail. The latter refers to the image itself.

kevintab95 replied:
Done.

seanlip commented on core/domain/topic_services.py:
```diff
@@ -299,12 +301,18 @@ def apply_change_list(topic_id, change_list):
                       topic_domain.TOPIC_PROPERTY_LANGUAGE_CODE):
                     topic.update_language_code(change.new_value)
+                elif (change.property_name ==
+                      topic_domain.TOPIC_PROPERTY_THUMBNAIL):
```
Similarly -- THUMBNAIL_FILENAME

kevintab95 replied:
Done.

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -51,6 +51,10 @@ class TopicModel(base_models.VersionedModel):
+    # The abbreviated name of the topic.
+    abbreviated_name = ndb.StringProperty(required=True, indexed=True)
+    # The thumbnail of the topic.
+    thumbnail = ndb.StringProperty(indexed=True)
```
thumbnail_filename (please fix the comment too)

kevintab95 replied:
Done.

seanlip commented on core/templates/dev/head/components/entity-creation-services/topic-creation.service.ts:
```diff
@@ -55,26 +63,28 @@ angular.module('oppia').factory('TopicCreationService', [
-        modalInstance.result.then(function(topicName) {
-          if (topicName === '') {
+        modalInstance.result.then(function(topic) {
+          if (topic.topicName === '') {
```
I don't see a warning here for empty abbreviated topic name?

kevintab95 replied:
Done.

seanlip commented on core/templates/dev/head/domain/topic/topic-update.service.ts:
```diff
@@ -124,6 +126,44 @@ angular.module('oppia').factory('TopicUpdateService', [
+       * undo/redo service.
+       */
+      setThumbnail: function(topic, thumbnailName) {
+        var oldThumbnailName = angular.copy(topic.getThumbnail());
```
Again, be careful with the terminology -- distinguish between "thumbnail" and "thumbnail filename". Also I'm not sure "thumbnail name" is a thing...

kevintab95 replied:
done,

seanlip commented on core/tests/protractor_utils/TopicsAndSkillsDashboardPage.js:
```diff
@@ -125,13 +126,14 @@ var TopicsAndSkillsDashboardPage = function() {
   };
 
-  this.createTopicWithTitle = function(title) {
+  this.createTopicWithTitle = function(title, shortName) {
```
Use consistent terminology throughout the codebase -- either shortName or abbreviatedName, but not both.

kevintab95 replied:
Done.

seanlip reviewed (changes requested):
Looks great! Just minor comments.

seanlip commented:
Also please ensure that, in normal operation, errors (such as abbreviated topic name being too long) don't even make it to the backend.

Commit:
address comments

Commit:
Merge remote-tracking branch 'upstream/develop' into abbreviated-topic-name-and-thumbnail

seanlip commented on core/templates/dev/head/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.ts:
```diff
@@ -83,7 +85,7 @@ angular.module('oppia').directive('topicEditorTab', [
           $scope.getStaticImageUrl = UrlInterpolationService.getStaticImageUrl;
 
-          var saveTopicThumbnail = function(imageURI) {
+          var saveTopicThumbnailUrl = function(imageURI) {
```
Should this be saveTopicThumbnailImageData? Seems like the thing that's being saved is an "image data" thing.

Also is it an image URI or an image data URI? The former suggests it's a link to a resource, but if that's the case it's a bit weird to call convertImageDataToImageFile on it. Can you clarify what exactly the imageURI represents?

kevintab95 replied:
Hi @seanlip, it is an image data URI. I'll rename it to saveTopicThumbnailImageData. Thanks.

seanlip commented on core/tests/protractor_utils/TopicsAndSkillsDashboardPage.js:
```diff
@@ -125,13 +126,14 @@ var TopicsAndSkillsDashboardPage = function() {
   };
 
-  this.createTopicWithTitle = function(title) {
+  this.createTopicWithTitle = function(title, abbreviatedName) {
```
Should this be renamed to this.createTopicWithTitleAndAbbreviatedName (or possibly just this.createTopic)?

kevintab95 replied:
Changed this to createTopic. Thanks!

seanlip reviewed (changes requested):
Thanks @kevintab95! Just a couple notes.

Commit:
fixes tests; addresses comments

kevintab95 commented:
Noticed an issue when editing/updating the thumbnail after the changes related to renaming. Working on it.

seanlip commented:
Thanks! Sounds good. Please note the backend coverage diff gap, too.

Also, please could you respond to all my comments? Otherwise it's hard to tell what's been reviewed and what's not.

Thanks!

aks681 commented on core/domain/topic_domain.py:
```diff
@@ -421,6 +431,31 @@ def __init__(
         self.story_reference_schema_version = story_reference_schema_version
 
+    @classmethod
+    def from_dict(cls, topic_dict):
```
Oh, sorry, there was no need to add this function. I thought this function already existed and just wanted to check if this was updated. Since it didn't exist before, and if you are not using this anywhere else, you can remove it.

kevintab95 replied:
Done.

aks681 commented on core/templates/dev/head/domain/topic/topic-update.service.ts:
```diff
@@ -70,7 +72,7 @@ angular.module('oppia').factory('TopicUpdateService', [
         property_name: propertyName,
         new_value: angular.copy(newValue),
-        old_value: angular.copy(oldValue)
+        old_value: angular.copy(oldValue) || null
```
Why this change? `oldValue` should never be  undefined, right?

kevintab95 replied:
This is to support topics that were created before these changes go in. Since older topics may not have oldValue (ie. it will be undefined) for abbreviated name and thumbnail fields, the changelist JSON sent to the backend might not contain old_value key and would throw a backend error.

Commit:
fix issue with editing thumbnail

Commit:
address comments

Commit:
fix tests

Commit:
Merge remote-tracking branch 'upstream/develop' into abbreviated-topic-name-and-thumbnail

aks681 commented:
I pulled this branch and when I removed the `required=True` from `abbreviated_topic_name`, a topic created on develop was loaded here with no errors! 
Some issues that I noticed:
- I can't seem to even Save Changes for the topic without uploading thumbnail. This is not expected, right? I thought only publishing was prevented.

- The name of the image is not modified on upload, so when I upload the same image again, an error is thrown. This shouldn't be the case. Maybe append time or randomize the name, to make the image name unique?

Other than that, I think it looks good!

kevintab95 commented:
Awesome thanks @aks681 for testing this. :)

In that case I can remove `required=True` from abbreviated_name so that it does not break any thing on develop.

I'll add a separate validation for pre-publish issues so that Save Changes works for the topic without uploading thumbnail and will make the filename unique like you suggested. Thanks!

Commit:
address comments

Commit:
fix lint issue

Commit:
fix frontend test failure

Commit:
change topicAndStoryEditor to run in dev mode

kevintab95 commented:
Hi @seanlip and @aks681, I've addressed the review comments PTAL. Thanks!

aks681 reviewed (approved):
LGTM! I checked the UI and it seems to be working fine :)

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -52,7 +52,7 @@ class TopicModel(base_models.VersionedModel):
     canonical_name = ndb.StringProperty(required=True, indexed=True)
     # The abbreviated name of the topic.
-    abbreviated_name = ndb.StringProperty(required=True, indexed=True)
+    abbreviated_name = ndb.StringProperty(indexed=True)
```
Could you please (a) add an audit job to check that all topics have a non-empty abbreviated name, and (b) file an issue to make this a required field once this job runs successfully?

kevintab95 replied:
Hi @seanlip, regarding (a), should the job also populate the abbreviated name field (eg. copy over the topic name?) if it is found to be empty?

seanlip replied:
No, it shouldn't. The flow should be that, in the release after this one, we run this just before a new deployment (and repeat until all the old abbreviated names are non-empty), and only then, do the new deployment.

Please could you make sure to update this form as well, with clear instructions, in order to ensure this is handled correctly? https://docs.google.com/forms/d/e/1FAIpQLSfvYWscAn18ok06An1oQ54h1VmBHfCX8uuuV01kIvY9WX0-Ug/viewform Alternatively you can email @ankita240796, @DubeySandeep and me with full instructions.

Thanks!

kevintab95 replied:
Added an audit test for the same.

seanlip commented on core/templates/dev/head/pages/topic-editor-page/editor-tab/topic-editor-tab.directive.ts:
```diff
@@ -141,6 +141,21 @@ angular.module('oppia').directive('topicEditorTab', [
             });
           };
 
+          var generateImageFilename = function() {
```
I feel like I've seen this logic elsewhere in the codebase before. Is there a way to share code rather than duplicate it?

kevintab95 replied:
Done.

seanlip reviewed (changes requested):
Thanks @kevintab95! Couple small notes.

seanlip commented:
Adding @vojtechjelinek @vinitamurthi @nithusha21 @ankita240796 @DubeySandeep @sophiewu6  for codeowner reviews.

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks!

ankita240796 reviewed (approved):
LGTM from code-owner's perspective,  Thanks @kevintab95!

nithusha21 reviewed (approved):
LGTM as a codeowner!

aks681 commented:
@kevintab95 One thing, I think the image, right now, is restricted to a square 1:1 aspect ratio? I think for the Android app, we need a 16:9 thumbnail. So, can you change that as well?

kevintab95 commented:
@aks681 I could remove the restriction of square shaped images and use the image uploader that we use in the editor page. Let me know what you think. Thanks!

aks681 commented:
We do need a restriction, just that it should be cropped to 16:9 instead of 1:1. Would that be possible relatively quickly? If not, you can create a separate issue and do it in a separate PR, maybe.

kevintab95 commented:
Oh Okay. I do not think the [library](https://github.com/alexk111/ngImgCrop) we use currently allows that flexibility. The user will have to create 16:9 image with padding such that it fits a square with the current approach. I think it is better to create a separate issue for this.

Commit:
address review comments; add audit job to check for empty abbreviated name fields and test

DubeySandeep commented on core/templates/dev/head/services/assets-backend-api.service.ts:
```diff
@@ -198,9 +202,16 @@ angular.module('oppia').factory('AssetsBackendApiService', [
+      } else if (assetType === ASSET_TYPE_THUMBNAIL) {
+        urlTemplate = THUMBNAIL_DOWNLOAD_URL_TEMPLATE;
+      } else {
+        urlTemplate = IMAGE_DOWNLOAD_URL_TEMPLATE;
```
Should we use a dict ASSET_TYPE_TO_DOWNLAOD_URL_TEMPLATE?

(Considering I'm going to add another assetType for voiceover applications.)

kevintab95 replied:
Done.

DubeySandeep commented on core/templates/dev/head/services/assets-backend-api.service.ts:
```diff
@@ -279,6 +290,10 @@ angular.module('oppia').factory('AssetsBackendApiService', [
         return _getDownloadUrl(
           entityType, entityId, filename, ASSET_TYPE_IMAGE);
+      },
+      getTopicThumbnailUrlForPreview: function(entityType, entityId, filename) {
```
This service should be unaware of "Topic" word, so I would suggest you use `getThumbnailUrlForPreview`.

kevintab95 replied:
Done.

DubeySandeep reviewed (approved):
LGTM for the codeowner files. I've left 2 minor comments PTAL.

seanlip commented on core/domain/prod_validation_jobs_one_off.py:
```diff
@@ -3531,12 +3531,25 @@ def _validate_uncategorized_skill_ids_not_in_subtopic_skill_ids(cls, item):
+        Args:
+            item: ndb.Model. TopicModel to validate.
+        """
+        if item.abbreviated_name == '' or item.abbreviated_name is None:
```
Maybe just "if not item.abbreviated_name"?

kevintab95 replied:
Done.

seanlip commented on extensions/objects/templates/filepath-editor.directive.ts:
```diff
@@ -794,17 +794,8 @@ angular.module('oppia').directive('filepathEditor', [
-            ('0' + date.getSeconds()).slice(-2) +
-            '_' +
-            Math.random().toString(36).substr(2, 10) +
+          var imageFilename = ImageUploadHelperService.generateImageFilename();
```
Hm, one thought. Do you think we should add height/width information to the topic thumbnail image too, similar to what we do for the images in lessons?

The reason we originally did that was because, if an image takes a while to load, then knowing its intended dimensions means that we can avoid the page juddering once it loads. It seems to me like this might be the case for all images.

kevintab95 replied:
Done. Used 150x150 now for topic thumbnail but this will be changed when we allow the cropping tool to crop at 16:9 aspect ratio.

Commit:
address comments

Commit:
address remaining comments

seanlip commented on core/storage/topic/gae_models.py:
```diff
@@ -52,7 +52,7 @@ class TopicModel(base_models.VersionedModel):
     canonical_name = ndb.StringProperty(required=True, indexed=True)
     # The abbreviated name of the topic.
-    abbreviated_name = ndb.StringProperty(indexed=True)
+    abbreviated_name = ndb.StringProperty(indexed=True, default="")
```
By default, we use single quotes in Python.

kevintab95 replied:
Done.

seanlip commented on core/templates/dev/head/services/image-upload-helper.service.ts:
```diff
@@ -52,7 +52,7 @@ angular.module('oppia').factory('ImageUploadHelperService', [
       },
 
-      generateImageFilename: function() {
+      generateImageFilename: function(height, width) {
```
Probably cleaner to pass fileExtension in as well, since that's part of the filename.

There should probably be unit tests written for this function too?

kevintab95 replied:
Done.

Commit:
address review comments

seanlip reviewed (approved):
Thanks! LGTM.

seanlip commented:
Well, almost -- you might need to merge develop into your branch to handle the coverage issues.

@kevintab95 could you please also file an issue to use the correct aspect ratio for the thumbnail image, so we don't forget about it?

Also, @DubeySandeep @vinitamurthi @sophiewu6 PTAL (as code owners). Thanks!

Commit:
Merge remote-tracking branch 'upstream/develop' into abbreviated-topic-name-and-thumbnail

kevintab95 commented:
Hi @seanlip, #8199 is the issue to correct thumbnail aspect ratio.

Commit:
Merge remote-tracking branch 'upstream/develop' into abbreviated-topic-name-and-thumbnail
